### ISSUE TITLE
[bit-titan] Found no results while trying to browse this tracker (Test)

### ISSUE THREAD

**Issue by Winblackka** [Jackett/Jackett#10281]
### Description

They have changed their search system. :(

### Logged Error Messages

System.Exception: Found no results while trying to browse this tracker
   at Jackett.Common.Services.IndexerManagerService.TestIndexer(String name) in /home/vsts/work/1/s/src/Jackett.Common/Services/IndexerManagerService.cs:line 282
   at Jackett.Server.Controllers.IndexerApiController.Test() in /home/vsts/work/1/s/src/Jackett.Server/Controllers/IndexerApiController.cs:line 130
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Jackett.Server.Middleware.CustomExceptionHandler.Invoke(HttpContext httpContext) in /home/vsts/work/1/s/src/Jackett.Server/Middleware/CustomExceptionHandler.cs:line 61

**Comment by garfield69** [comment]
yup. no problem. I'll adjust the indexer to work with the new search.

**Comment by Winblackka** [comment]
Wow, super fast, thx!

**Comment by Winblackka** [comment]
Do you need logs or something else?

**Comment by garfield69** [comment]
I've coded the indexer, but I've run into a snag, and am waiting for our resident C# dev to take a look.
The site's new ajax is causing the Cardigann processor to attempt to re-login **after** fetching the results, and it fails because the site uses captcha. I cannot figure out why there is a re-login. :-(

**Comment by Winblackka** [comment]
Oh dammit, I hope you'll find a working method for it.

**Comment by T89-BT** [comment]
Hey guys, I'm the developer of BiT-TiTAN, it is possible to provide you with an API. This is definitely easier for you and much better for us!

**Comment by ngosang** [comment]
I don't have time to migrate from Cardignann to C# this indexer. We need some volunteer, I can review the code, etc...
@T89-BT if you can provide an API the integration will be easy. We need the following requirements:
* Login endpoint (or support some king of token)
* Search endpoint
  * Support for text search in the title
  * Support for text search + categories
  * Support for empty text search => return 100 latests torrents

**Comment by T89-BT** [comment]
ok that's not a problem ... the api gives as response json, xml or html. Which parameters are required as a response? are name, upload date, size and category sufficient or are additional parameters required

**Comment by ngosang** [comment]
The columns in this screenshot. Include the detail page link and the download link. DL and UL factors are related to freeleech, doble upload, etc... We also support imdb search + imdb id in the response but it's optional.
![image](https://user-images.githubusercontent.com/10577978/100542473-59c54000-324a-11eb-8502-4964cfba18bb.png)

**Comment by T89-BT** [comment]
API Response via XML

`<btxml>
<siteUrl>https://bit-titan.net/</siteUrl>
<siteName>BiT-TiTAN</siteName>
<filesFound>39020</filesFound>
<pages>1561</pages>
<viewResults>0 to 25 from 39020</viewResults>
<results>
<file>
<id>383774</id>
<name>Typisch.Andy!.(2001).S01-S03.SD.TVrip</name>
<size>10697264354</size>
<category>2080</category>
<seeds>0</seeds>
<leechers>0</leechers>
<snatchers>0</snatchers>
<downloadFactor>0</downloadFactor>
<uploadFactor>1</uploadFactor>
<categoryName>Serien - SD Packs</categoryName>
<download>https://bit-titan.net/api.php?download=383774&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv</download>
<added>Mon, 23 Nov 2020 1:26:16 +0100</added>
</file>
<file>
<id>383773</id>
<name>Himmelstal.S01.German.DL.1080p.BluRay.x264-ROCKEFELLER</name>
<size>22343980796</size>
<category>2070</category>
<seeds>0</seeds>
<leechers>0</leechers>
<snatchers>0</snatchers>
<downloadFactor>0</downloadFactor>
<uploadFactor>1</uploadFactor>
<categoryName>Serien - HD Packs</categoryName>
<download>https://bit-titan.net/api.php?download=383773&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv</download>
<added>Mon, 23 Nov 2020 1:03:54 +0100</added>
</file>
<file>
<id>383772</id>
<name>Trust.S01E04.German.DL.720p.WEB.h264-WvF</name>
<size>2172542314</size>
<category>2040</category>
<seeds>0</seeds>
<leechers>0</leechers>
<snatchers>0</snatchers>
<downloadFactor>0</downloadFactor>
<uploadFactor>1</uploadFactor>
<categoryName>Serien - x264</categoryName>
<download>https://bit-titan.net/api.php?download=383772&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv</download>
<added>Mon, 23 Nov 2020 0:47:59 +0100</added>
</file>
</btxml>`

Response via JSON

`{"siteUrl":"https:\/\/bit-titan.net\/","siteName":"BiT-TiTAN","filesFound":"39020","pages":1561,"viewResults":"0 to 25 from 39020","results":[{"id":"383774","name":"Typisch.Andy!.(2001).S01-S03.SD.TVrip","size":"10697264354","category":"2080","seeds":"0","leechers":"0","snatchers":"0","downloadFactor":"0","uploadFactor":"1","categoryName":"Serien - SD Packs","download":"https:\/\/bit-titan.net\/api.php?download=383774&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv","added":"Mon, 23 Nov 2020 1:26:16 +0100"},{"id":"383773","name":"Himmelstal.S01.German.DL.1080p.BluRay.x264-ROCKEFELLER","size":"22343980796","category":"2070","seeds":"0","leechers":"0","snatchers":"0","downloadFactor":"0","uploadFactor":"1","categoryName":"Serien - HD Packs","download":"https:\/\/bit-titan.net\/api.php?download=383773&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv","added":"Mon, 23 Nov 2020 1:03:54 +0100"}]}`



Response via HTML

`<!DOCTYPE html>
--
  | <html>
  | <head>
  | <title>BiT-TiTAN-API</title>
  | </head>
  | <body>
  | <ul>
  | <li class="siteUrl">https://bit-titan.net/</li>
  | <li class="siteName">BiT-TiTAN</li>
  | <li class="filesFound">39020</li>
  | <li class="pages">1561</li>
  | <li class="viewResults">0 to 25 from 39020</li>
  | <li class="results">
  | <ul class="file">
  | <li class="id">383774</li>
  | <li class="name">Typisch.Andy!.(2001).S01-S03.SD.TVrip</li>
  | <li class="size">10697264354</li>
  | <li class="category">2080</li>
  | <li class="seeds">0</li>
  | <li class="leechers">0</li>
  | <li class="snatchers">0</li>
  | <li class="downloadFactor">0</li>
  | <li class="uploadFactor">1</li>
  | <li class="categoryName">Serien - SD Packs</li>
  | <li class="download">https://bit-titan.net/api.php?download=383774&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv</li>
  | <li class="added">Mon, 23 Nov 2020 1:26:16 +0100</li>
  | </ul>
  | <ul class="file">
  | <li class="id">383773</li>
  | <li class="name">Himmelstal.S01.German.DL.1080p.BluRay.x264-ROCKEFELLER</li>
  | <li class="size">22343980796</li>
  | <li class="category">2070</li>
  | <li class="seeds">0</li>
  | <li class="leechers">0</li>
  | <li class="snatchers">0</li>
  | <li class="downloadFactor">0</li>
  | <li class="uploadFactor">1</li>
  | <li class="categoryName">Serien - HD Packs</li>
  | <li class="download">https://bit-titan.net/api.php?download=383773&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv</li>
  | <li class="added">Mon, 23 Nov 2020 1:03:54 +0100</li>
  | </ul>
  | <ul class="file">
  | <li class="id">383772</li>
  | <li class="name">Trust.S01E04.German.DL.720p.WEB.h264-WvF</li>
  | <li class="size">2172542314</li>
  | <li class="category">2040</li>
  | <li class="seeds">0</li>
  | <li class="leechers">0</li>
  | <li class="snatchers">0</li>
  | <li class="downloadFactor">0</li>
  | <li class="uploadFactor">1</li>
  | <li class="categoryName">Serien - x264</li>
  | <li class="download">https://bit-titan.net/api.php?download=383772&apiKey=JClceuKdMlda1EMYcXjsRUfEt7e2mEAv</li>
  | <li class="added">Mon, 23 Nov 2020 0:47:59 +0100</li>
  | </ul>
</li>
--
  | </ul>
  | </body>
  | </html>
`



Default acces via https:/bit-titannet/api.php?apiKey=xXxAPIKEYxXx&search=&categories=1001,1002,1003&searchIn=0&uploaderID=2683&orderBy=1&order=1&limit=1&page=1&response=1&timeFormat=1&sizeFormat=1&categoryName=1&downloadLink=1&


I hope you can work with it .... I think it will go online in the course of the week

**Comment by Winblackka** [comment]
Any new results? Is somebody working on it? :)

**Comment by hosoludi** [comment]
Would be nice have this working again.

**Comment by T89-BT** [comment]
The API is now online with us. You can find the control panel at https://bit-titan.net/api_cp.php

There you can then generate your API key and an API link.

Default Link: https://bit-titan.net/api.php?apiKey=xxxXXXxxxXXXxxxXXXxxxXXXxxxXXXxxx&search=&categories=0&searchIn=0&uploaderID=2683&orderBy=1&order=1&limit=1&page=1&response=1&timeFormat=1&sizeFormat=1&categoryName=0&downloadLink=0

Parameters for Response

- apiKey -> is required an musst generate in the cp

- search -> enter a search string if you want (allowed chars a-z A-Z 0-9 ._-+()[]) other chars replace the script

- categories -> 0 for disabled or category, more categorys separate with a comma (1010,1020,1030)

- searchIn -> 
    0 = All Files (DEFAULT VALUE)
    1 = active Torrents (min. 1 seeder)
    2 = died Torrents
    3 = cinema highlights
    4 = own uploads or set a uploaderID for his uploads
    5 = only uploads by bots
    6 = only OnlyUpload torrents
    7 = only multiplikator torrents
    8 = only torrents with percentage download
    9 = only freeleech torrents

- uploaderID -> if you set searchIn to 4 you can set a uploader ID

- orderBy ->
   1 = sorted by upload date (DEFAULT VALUE)
   2 = sorted by number of seeders
   3 = sorted by number of leechers
   4 = sorted by number of snatchers
   5 = sorted by file size
   6 = sorted by number of comments
   7 = sorted by categorie ID
   8 = sorted by number of files
   9 = sorted by torrents name
 10 = sorted by number of torrent hits
 11 = sorted by uploader ID

- order ->
   1 = DESC (DEFAULT VALUE)
   2 = ASC

- limit ->
    1 = get 25 items (DEFAULT VALUE)
    2 = get 50 items
    3 = get 75 items
    4 = get 100 items

- page -> set the page of results (DEFAULT IS 1)

- response ->
    1 = JSON response (DEFAULT VALUE)
    2 = XML response
    3 = HTML response

- timeFormat ->
    1 = Sat, 28 Nov 2020 23:29:38 +0100 (DEFAULT VALUE)
    2 = 1606606178 (UNIX TIMESTAMP)
    3 = 28.11.2020 23:29.38 (German Timestamp)
    4 = 2020-11-28 23:29:38 (Default Datetime)

-  sizeFormat ->
    1 = outputs an integer value in bytes
    2 = outputs an string value in bytes, kilobytes, megabytes, usw (700 MB or 1,50GB)

- categoryName -> set 0 for False 1 for TRUE (if true, the category name is also output)

- downloadLink -> set 0 for False 1 for TRUE (if true, a download link is output)

**Comment by theonly112** [comment]
@T89-BT Thank you so much for providing an API.
I've been having a go at implementing an indexer for the API and noticed that if a limit is provided that is larger than the number of actual results, the API  does not return anything. For example: 73 Results are available. With "limit=3" nothing is returned. With "limit=2" the first 50 results are returned, and there is a second page available with the remaining 23.
I assume this is not done on purpose?

**Comment by T89-BT** [comment]
No, that would not be wanted.

I tried to simulate the error. However, I get my results even at Limit 4 (100 results)
with the following call:
https://bit-titan.net/api.php?apiKey=xxxXXXxxx&search=king.of.queens&categories=0&searchIn=0&&orderBy=1&order=1&limit=4&page=1&response=1&timeFormat=1&sizeFormat=1&categoryName=0&downloadLink=1&


If I generate exactly one result, this is the output:
{"siteUrl":"https:\/\/bit-titan.net\/","siteName":"BiT-TiTAN","filesFound":"1","pages":1,"viewResults":"0 to 100 from 1","results":[{"id":"294491","name":"King.of.Queens.Complete.German.AC3.DL.1080p.BluRay.x265-FuN","size":"108204221622","category":"2070","seeds":"2","leechers":"1","snatchers":"9","downloadFactor":"0","uploadFactor":"2","categoryName":"Serien - HD Packs","download":"https:\/\/bit-titan.net\/api.php?downloadID=294491&apiKey=xxxXXXxxx","added":"Wed, 01 Jul 2020 19:41:19 +0100"}]}

Can you maybe send me an API link as an example?



The API does not currently search in Imdb titles or IDs and description, but this can be extended if desired

**Comment by theonly112** [comment]
@T89-BT 
Thanks for looking into it.
Here's an example:
https://bit-titan.net/api.php?apiKey=xxxXXXxxx&search=Bond&downloadLink=1&timeFormat=1&page=2&limit=1
Returns 25 to 50 (of 71 total)
https://bit-titan.net/api.php?apiKey=xxxXXXxxx&search=Bond&downloadLink=1&timeFormat=1&page=3&limit=1
Returns nothing at.
Am I using the API incorrectly?

Search by IMDB Id would be great for services like Radarr.

**Comment by T89-BT** [comment]
No, the use of the API is correct. On the last page there was an error with UTF-8 encoding in the JSON array, I have already fixed this and will update it to BT later.

But you can also set the limit to 4, then you immediately have 100 entries and don't have to flip through the pages

**Comment by garfield69** [comment]
@T89-BT  I do not get any results when using the API with a search for corona or for category 9020
https://bit-titan.net/api.php?apiKey=myapikey&categories=0&search=corona&downloadLink=1&limit=4
https://bit-titan.net/api.php?apiKey=myapikey&categories=9020&search=&downloadLink=1&limit=4
both return a http 200 but with no response data.
the web site returns results.
Could you investigate please? Thanks.

**Comment by garfield69** [comment]
v0.18.805

**Comment by montana123** [comment]
If you put &limit=4 and no search Term the API returns empty http 200 Object.

@garfield69 Maybe leave limit to 1? Is this possible or is this not good?

How about the Page Settings? I will upload a new pull request because currently bit-titan.cs is bugged because of limit=4 in their api

**Comment by garfield69** [comment]
um, not liking having to fetch multiple pages of 25 results each, in an attempt to fill the results list up to 100.
the bit-titan docs state that the &limit=4 should return up to 100 results per page and therefore we should only need to fetch one page.

I would rather the developers on Bit-Titan fix their API bug.

BTW your PR12351 code is setting &page1 which is not valid. should be &page=1.
You are currently fetching four times the first 25 results presenting 100 results of 4x duplicates ;-D

But even with the correction I cannot get more than two pages returned, despite there apparently being more available.

`https://bit-titan.net/api.php?apiKey=myapikey&search=&downloadLink=1&limit=1&page=1`
returns
WebClient(HttpWebClient2): Returning OK => 8960 bytes
```json
{"siteUrl":"https:\/\/bit-titan.net\/","siteName":"BiT-TiTAN","filesFound":"87106","pages":3485,"viewResults":"0 to 25 from 87106","results":[{"id":"646185" ... etc etc etc
```
BTW `viewResults":"0 to 25 from 87106"` should say `0-to 24`

second page
`https://bit-titan.net/api.php?apiKey=myapikey&search=&downloadLink=1&limit=1&page=2`
returns 
WebClient(HttpWebClient2): Returning OK => 8945 bytes
```json
{"siteUrl":"https:\/\/bit-titan.net\/","siteName":"BiT-TiTAN","filesFound":"87106","pages":3485,"viewResults":"25 to 50 from 87106","results":[{"id":"646160" ... etc etc etc
```

third page
`https://bit-titan.net/api.php?apiKey=myapikey&search=&downloadLink=1&limit=1&page=3`
returns
WebClient(HttpWebClient2): Returning OK => 0 bytes
Error on 20210930064812 for BiT-TiTAN.txtSystem.NullReferenceException: Object reference not set to an instance of an object.

I've also tried the combo of limit=2 and fetching 2 pages, I get one page of 50 then 0 bytes for the page=2 request.
and using limit=3 or limit=4 both return 0 bytes, which I am fairly certain was working previously when I tested my limit=4 update originally :-(

So in conclusion we are currently restricted to a max of 50 results, using either two requests for page=1 and page=2 with limit=1 
 for a total of 50 results
or
one request of page=1 with limit=2 for 50 results

so I am going to decline your PR and just set a limit=2, which in my tests so far appears to be working for now.
And hope the Bit-Titan developers will fix this and the earlier reported but about category 9020 and searching for corona.

**Comment by garfield69** [comment]
the api is returning http 200 but with empty response

**Comment by garfield69** [comment]
never mind. I logged out of the web site and tried the indexer again and now its working.

**Comment by garfield69** [comment]
it's doing it again, the api is returning http 200 but with empty response

**Comment by garfield69** [comment]
On the browser I changed the api key, and logged out and back in, even deleted the browser session cookies, but the api page example URI that the page generates for API testing does not return any response other than the http 200.
Same with Jackett :-(
Just me or is the API broken?

**Comment by garfield69** [comment]
now working. buggered if I understand what is going on.
